### PR TITLE
Read and init runtime auth from config file

### PR DESCRIPTION
**Pull request by mitchdevenport**

### CODE REVIEW

File: bin/spiced/src/lib.rs [hunk 1]
```diff
@@ -42,6 +42,8 @@ pub struct Args {
 
 pub async fn run(args: Args) -> Result<()> {
     let app = App::new(".").context(UnableToConstructSpiceAppSnafu)?;
+    let auth = runtime::auth::get();
```
Comment (phillipleblanc) [inline bin/spiced/src/lib.rs]: This is fine for now when Spice is the only provider - but we will need a way to load datasource specific auth based on which datasources are configured.

File: crates/runtime/src/auth.rs [hunk 2]
```diff
@@ -1,21 +1,55 @@
...
-    #[must_use]
-    pub fn get_auth(name: &str) -> Box<dyn Auth> {
-        match name {
-            "spiceai" => {
-                todo!("SpiceAI auth not implemented yet")
-            }
-            _ => Box::new(none::NoneAuth::new()),
+#[must_use]
+pub fn get() -> Box<dyn Auth> {
+    let auth_path = match dirs::home_dir() {
```
Comment (phillipleblanc) [inline crates/runtime/src/auth.rs]: I don't think it makes sense for the file reading logic to happen here - this is specific to the spice auth, and so I would expect it to be in the Spice auth implementation

Comment (phillipleblanc) [reply crates/runtime/src/auth.rs]: How would this be extended to support another auth provider?

File: crates/runtime/src/auth.rs [hunk 3]
```diff
@@ -1,21 +1,55 @@
...
-impl AuthProviders {
-    #[must_use]
-    pub fn get_auth(name: &str) -> Box<dyn Auth> {
-        match name {
-            "spiceai" => {
-                todo!("SpiceAI auth not implemented yet")
-            }
-            _ => Box::new(none::NoneAuth::new()),
+#[must_use]
+pub fn get() -> Box<dyn Auth> {
```
Comment (phillipleblanc) [inline crates/runtime/src/auth.rs]: The reason I had the interface as `get_auth(name: &str)` was that I was expecting the runtime to load specific auth providers depending on which data sources needed it. Not a single auth object

File: bin/spiced/src/lib.rs [hunk 4]
```diff
@@ -42,6 +42,18 @@ pub struct Args {
...
+    match auth.parse_from_config() {
+        Ok(()) => {}
+        Err(e) => {
+            tracing::error!(
+                "Unable to parse auth from config, proceeding without auth: {}",
+                e
+            );
+        }
+    }
+    tracing::info!("Spice Auth token: {}", auth.get("spice.ai").get_token());
```
Comment (sgrebnov) [inline bin/spiced/src/lib.rs:55]: @mitchdevenport  -  do we plan a scenario when we will want to use several providers in the same time so we would like to have `auth.get("spice.ai").get_token()` (as it is proposed here) vs single auth provider instance so it is better to do `auth = runtime::auth::AuthProviders::get("spice.ai")` and then just 'auth.get_token()'


Is it a universal auth we will use to communicate with different services so need support for multiple providers for the same auth instance?

**Review by sgrebnov — APPROVED** [review]

File: bin/spice/pkg/api/auth.go [hunk 5]
```diff
@@ -1,9 +1,10 @@
 package api
 
-type AuthConfig struct {
-	Auth *Auth `json:"auth,omitempty" csv:"auth" toml:"auth"`
-}
+const (
+	AUTH_TYPE_SPICE_AI = "spice.ai"
```
Comment (phillipleblanc) [inline bin/spice/pkg/api/auth.go:4]: This should be an Enum?

File: crates/runtime/src/auth.rs [hunk 6]
```diff
@@ -1,21 +1,80 @@
...
 impl AuthProviders {
     #[must_use]
-    pub fn get_auth(name: &str) -> Box<dyn Auth> {
-        match name {
-            "spiceai" => {
-                todo!("SpiceAI auth not implemented yet")
-            }
+    pub fn get(&self, name: &str) -> Box<dyn AuthProvider> {
+        let auth_provider = if let Some(auth_provider) = self.auth.get(name) {
+            tracing::info!("Using auth provider: {}", auth_provider.provider_type);
```
Comment (phillipleblanc) [inline crates/runtime/src/auth.rs:53]: Let's move these to `trace!` level, not `info!`

File: crates/runtime/src/auth.rs [hunk 7]
```diff
@@ -1,21 +1,80 @@
...
+            auth_provider
+        } else {
+            tracing::info!("No auth provider found for {}", name);
+            return Box::new(none::NoneAuth::new());
+        };
+
+        match auth_provider.provider_type.as_str() {
```
Comment (phillipleblanc) [inline crates/runtime/src/auth.rs:60]: This should be matching on a Enum

File: bin/spice/pkg/api/auth.go (hunk 5 above)
Comment (phillipleblanc) [reply bin/spice/pkg/api/auth.go:4]: Ah this is Go code

**Pull request merged**